frontend/apps/reader/modules/readeruserhyph.lua

```diff
@@ -0,0 +1,321 @@
+        self.dict = io.open(self.dict_file, "w")
+        self.dict:close()
+    end
+
+    -- open files
+    self.dict = io.open(self.dict_file, "r")
+    if not self.dict then
+        logger.err("UserHyph: could not open " .. self.dict_file)
+        return
+    end
+    self.new_dict = io.open(self.new_dict_file, "w")
+    if not self.new_dict then
+        logger.err("UserHyph: could not open " .. self.new_dict_file)
+        return
+    end
+    return true
+end
+
+-- Reads the first dictionary entries from the old dictionary
+-- and writes them to the new dict, as long the entry is alphabetically lower than (_case insensitive_) word
+-- the last line is not written, but returned as: word, line
+--
+-- the format is line per line: "word;hyphenation\n"
+-- e.g.: "danger;dan-ger\n"
+function ReaderUserHyph:findEntry(word)
+    -- scan hyphenation dictionary for selected word
+    local word_lower = self.ui.document:getLowercasedWord(word)
+    local line = self.dict:read()
+    while line and self.ui.document:getLowercasedWord(line:sub(1, line:find(";") - 1)) < word_lower do
+        self.new_dict:write(line .. "\n")
+        line = self.dict:read()
+    end
+
+    if not line then -- EOF
+        return
+    end
+
+    return self.ui.document:getLowercasedWord(line:sub(1, line:find(";") - 1)), line
+
+    --[[ -- hyphenation from dictionary, not needed
+    local hyphenation
+    -- check if a hyphenation is found for word
+    if line and self.ui.document:getLowercasedWord(line:sub(1, line:find(";") - 1)) == word_lower then
+        hyphenation = string.sub(line, line:find(";") + 1) -- hyphenation found
+        line = nil -- Important for not duplicating the entry in the dictionary
+    end
+    return hyphenation
+    ]]
+end
+
+-- writes one entry to file
+function ReaderUserHyph:writeEntry(line)
+    if line then
+        self.new_dict:write(line .. "\n")
+    end
+end
+
+-- reads the rest of the old dictionary and writes this to the reset
+-- if line~=nil, an additional line can be inserted first
+function ReaderUserHyph:writeRest(line)
+    -- write old entry if there was one
+    if line then
+        self.new_dict:write(line .. "\n")
+    end
+    -- copy rest of file
+    repeat
+        line = self.dict:read()
+        if line then
+            self.new_dict:write(line .. "\n")
+        end
+    until (not line)
+end
+
+-- closes all open files and invalidates variables
+function ReaderUserHyph:closeDictionary()
+    self.dict:close()
+    self.new_dict:close()
+    os.remove(self.dict_file)
+    os.rename(self.new_dict_file, self.dict_file)
+
+    self.dict_file = nil
+    self.new_dict_file = nil
+    self.dict = nil
+end
+
+function ReaderUserHyph:modifyUserEntry(word)
+    if word:find("[ ,;-%.]") then return end -- no button if more than one word
+
+    if not self.ui.document then
+        return
+    end
+
+    local suggested_hyphenation = self.ui.document:getHyphenationForWord(word)
```

Comment:
Do I get it right? : you no longer need to open the dict file before cancel/update, because now you just ask crengine for it (as it has already loaded/parsed that file) ?
Nice - if I am right :)